### ISSUE TITLE
``services.redmine.settings.production.email_delivery.smtp_settings`` has no effect

### ISSUE THREAD

**Issue by massimo-zaniboni** [NixOS/nixpkgs#106992]
**Describe the bug**

``services.redmine.settings.production.email_delivery.smtp_settings`` and ``services.redmine.settings.production.email_delivery.sendmail_settings`` have no effect. 

**Expected behavior**

I had to configure equivalent options inside ``services.redmine.extraEnv``, despite the best place is ``services.redmine.settings``.

**To Reproduce**

In this configuration example I comment with ``GOOD`` and ``BAD`` 

```
services.redmine = {
  enable = true;
  # package = redmineCustom;
  database.name = "redmine";
  database.type = "postgresql";
  database.user = "redmine";
  database.port = 9054;

  port = 3000;

  settings = {
    production = {
      email_delivery = {

        # GOOD: this is read and activated
        delivery_method =  "smtp";

        # BAD: these settings does not make effect,
        # despite they are reported inside
        # > /var/lib/redmine/config/configuration.yml
        #
        smtp_settings = {
          enable_starttls_auto = true;
          address = "mail.example.com";
          port = 587;
          domain = "example.com";
          authentication = "plain";
          user_name = "test@example.com";
          password = "${pswd}";
        };
      };
    };
  };

  # GOOD: these settings are read and activated,
  # but this is a work around 
  extraEnv = ''
config.action_mailer.perform_deliveries = true
config.action_mailer.raise_delivery_errors = true

config.action_mailer.delivery_method = :smtp
config.action_mailer.smtp_settings = {
  address:              'mail.example.com',
  port:                 587,
  domain:               'example.com',
  user_name:            'test@example.com',
  password:             '${pswd}',
  authentication:       'plain',
  enable_starttls_auto: true,
  openssl_verify_mode: 'none'
};
  '';
};
```
**Notify maintainers**

@aanderse

**Metadata**

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
These paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/ajidkqg53y0zvh0m5cjgqiw3pf9cvlf0-bash-interactive-4.4-p23-dev
copying path '/nix/store/ajidkqg53y0zvh0m5cjgqiw3pf9cvlf0-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.72, NixOS, 20.09.2238.b94726217f7 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.9`
 - channels(root): `"nixos-20.09.2238.b94726217f7, nixos-unstable-21.03pre244045.1179840f9a8"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: [services.redmine]
```

**Comment by aanderse** [comment]
@massimo-zaniboni I have tested this and the issue appears to be that `pkgs.format.yaml {}.generate` writes `YAML` files with everything quoted.

ping @Infinisil

In https://github.com/NixOS/nixpkgs/commit/dee97b8b44d549fba19b14d351b1492cb6714152#diff-f27d972cacce9ad1efc939816e54a99ac406ff571053fe23ba50a3a233836ef2 I switched `redmine` from using an `extraConfig` to a `settings` option, utilizing `pkgs.formats.yaml`. From my preliminary investigation it seems like quoted `YAML` doesn't work well with `redmine` (presumably the `ruby` `actionmailer` package). I'm guessing there is a significant difference between `"sendmail"` and `:sendmail`...

`YAML` that works with redmine:
```
default:
  email_delivery:
    delivery_method: :sendmail
    sendmail_settings:
      arguments: "-i"
      location: "/run/wrappers/bin/sendmail"
```

`YAML` that **doesn't** work with `redmine`:
```
{
  "default": {
    "email_delivery": {
      "delivery_method": "sendmail",
      "sendmail_settings": {
        "arguments": "-i",
        "location": "/run/wrappers/bin/sendmail",
      }
    }
  },
```

Any ideas on that? Maybe a `toYAML` function that has a boolean `quote` option to generate `YAML` without quotes? :thinking:

**Comment by massimo-zaniboni** [comment]
> Any ideas on that? Maybe a toYAML function that has a boolean quote option to generate YAML without quotes? thinking

Probably it is better improving `toYAML` and if a key like `sendmail_settings` does not contain any special character, then it is rendered without the surrounding `"..."`. I'm not an expert of Nix, but I can inspect the usage of `toYAML` and if it seems that it does not break anything, maybe make a PR.

In the mean-time I can improve the documentation of Redmine service, because the current work-around is fully functional.

> I'm guessing there is a significant difference between "sendmail" and :sendmail...

mmh... I'm not an expert of RoR, but it seems that `"sendmail"` is equivalent to `:sendmail` and it is working correctly. The same for `:plain` and similar flags.

**Comment by aanderse** [comment]
@massimo-zaniboni thank you for your efforts. Do you mind opening an issue upstream with either `redmine` (or possibly `actionmailer`) and seeing if they have any advice?

**Comment by Infinisil** [comment]
The fact that `:sendmail` has a `:` while the json-equavalent doesn't isn't the problem?

**Comment by aanderse** [comment]
I don't believe so, because I can get `:smtp` or `"smtp"` to both work (with minimal options tested). After poking through a little `redmine` code (**warning:** I don't know any `ruby`) I'm not convinced that is the issue... but I could most certainly be wrong on that.

**Comment by aanderse** [comment]
@massimo-zaniboni did you end up filing an issue or getting any advice from upstream?

**Comment by aanderse** [comment]
@massimo-zaniboni ping

**Comment by massimo-zaniboni** [comment]
@aanderse sorry up to date I have no time. The problem is documented: probably if someone search on Google will find the work-around. I have more important things to do, not having a work-around :-/

**Comment by aanderse** [comment]
@massimo-zaniboni understood. Thanks for the reply!

**Comment by aanderse** [comment]
In theory this was fixed by #133807. Thanks @lucasew :rocket:
